core/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java

```diff
@@ -265,6 +265,10 @@ protected Query doToQuery(QueryParseContext parseContext) throws IOException {
         addBooleanClauses(parseContext, booleanQuery, shouldClauses, BooleanClause.Occur.SHOULD);
         addBooleanClauses(parseContext, booleanQuery, filterClauses, BooleanClause.Occur.FILTER);
 
+        if (booleanQuery.clauses().isEmpty()) {
+            return new MatchAllDocsQuery();
+        }
```

Comment:
a few lines above we also have:

```
if (!hasClauses()) {
    return new MatchAllDocsQuery();
}
```

that if is against our builder clauses, for the java api I guess, while this one is for resulting lucene clauses, in case all of the parsed clasues where empty_query? Wonder if we can have a single if like that, maybe the last one alone would do it?